Bottom: 8315c09e2eae26949b3ae447b152c7908c111587
Top:    0ae674457a2da0bb9a43d860a6be1ea88988e029
Author: Robert Collins <robertc@robertcollins.net>
Date:   2013-02-17 16:12:38 +1300

Add a dedicated interface for controlling an in-progress TestRun.


---

diff --git a/NEWS b/NEWS
index 4fd368c..7290e1f 100644
--- a/NEWS
+++ b/NEWS
@@ -27,6 +27,12 @@ Improvements
   series of dicts describing a test. Useful for writing trivial stream
   analysers. (Robert Collins)
 
+* New support class ``TestControl`` which permits cancelling an in-progress
+  run. (Robert Collins)
+
+* New support class ``StreamFailfast`` which calls a ``TestControl`` instance
+  to abort the test run when a failure is detected. (Robert Collins)
+
 * New test support class ``testtools.testresult.doubles.StreamResult``, which
   captures all the StreamResult events. (Robert Collins)
 
diff --git a/doc/for-framework-folk.rst b/doc/for-framework-folk.rst
index d3b7c0f..c7b3904 100644
--- a/doc/for-framework-folk.rst
+++ b/doc/for-framework-folk.rst
@@ -155,7 +155,11 @@ TestControl
 
 Lastly we define the ``TestControl`` API which is used to provide the
 ``shouldStop`` and ``stop`` elements from ``TestResult``. Again, see the API
-documentation for ``testtools.TestControl``.
+documentation for ``testtools.TestControl``. ``TestControl`` can be paired with
+a ``StreamFailfast`` to trigger failing aborting a test run when a failure
+is observed. Aborting multiple workers in a distributed environment requires
+hooking whatever signalling mechanism the distributed environment has up to
+a ``TestControl`` in each worker process.
 
 StreamToDict
 ------------
diff --git a/testtools/__init__.py b/testtools/__init__.py
index b78bf53..d30af69 100644
--- a/testtools/__init__.py
+++ b/testtools/__init__.py
@@ -26,9 +26,11 @@ __all__ = [
     'skip',
     'skipIf',
     'skipUnless',
+    'StreamFailfast',
     'StreamResult',
     'StreamSummary',
     'StreamToDict',
+    'TestControl',
     'ThreadsafeForwardingResult',
     'try_import',
     'try_imports',
@@ -71,11 +73,13 @@ else:
         CopyStreamResult,
         ExtendedToOriginalDecorator,
         MultiTestResult,
+        StreamFailfast,
         StreamResult,
         StreamSummary,
         StreamToDict,
         Tagger,
         TestByTestResult,
+        TestControl,
         TestResult,
         TestResultDecorator,
         TextTestResult,
diff --git a/testtools/testresult/__init__.py b/testtools/testresult/__init__.py
index 29ee39d..435cb32 100644
--- a/testtools/testresult/__init__.py
+++ b/testtools/testresult/__init__.py
@@ -6,11 +6,13 @@ __all__ = [
     'CopyStreamResult',
     'ExtendedToOriginalDecorator',
     'MultiTestResult',
+    'StreamFailfast',
     'StreamResult',
     'StreamSummary',
     'StreamToDict',
     'Tagger',
     'TestByTestResult',
+    'TestControl',
     'TestResult',
     'TestResultDecorator',
     'TextTestResult',
@@ -21,11 +23,13 @@ from testtools.testresult.real import (
     CopyStreamResult,
     ExtendedToOriginalDecorator,
     MultiTestResult,
+    StreamFailfast,
     StreamResult,
     StreamSummary,
     StreamToDict,
     Tagger,
     TestByTestResult,
+    TestControl,
     TestResult,
     TestResultDecorator,
     TextTestResult,
diff --git a/testtools/testresult/real.py b/testtools/testresult/real.py
index 8c382d6..af0d1f0 100644
--- a/testtools/testresult/real.py
+++ b/testtools/testresult/real.py
@@ -6,10 +6,12 @@ __metaclass__ = type
 __all__ = [
     'ExtendedToOriginalDecorator',
     'MultiTestResult',
+    'StreamFailfast',
     'StreamResult',
     'StreamSummary',
     'StreamToDict',
     'Tagger',
+    'TestControl',
     'TestResult',
     'TestResultDecorator',
     'ThreadsafeForwardingResult',
@@ -417,6 +419,18 @@ class CopyStreamResult(StreamResult):
         domap(methodcaller('status', *args, **kwargs), self.targets)
 
 
+class StreamFailfast(StreamResult):
+    """Call the supplied callback if an error is seen in a stream."""
+
+    def __init__(self, callback):
+        self.callback = callback
+
+    def status(self, test_id, test_status, test_tags=None, runnable=True,
+        route_code=None, timestamp=None):
+        if test_status in ('uxsuccess', 'fail'):
+            self.callback()
+
+
 class StreamToDict(StreamResult):
     """A specialised StreamResult that emits a callback as tests complete.
 
@@ -597,6 +611,23 @@ class StreamSummary(StreamToDict):
         self.unexpectedSuccesses.append(case)
 
 
+class TestControl(object):
+    """Controls a running test run, allowing it to be interrupted.
+    
+    :attribute shouldStop: If True, tests should no run and should instead
+        return without running. Similarly a TestSuite should abort running
+        and return immediately.
+    """
+
+    def __init__(self):
+        super(TestControl, self).__init__()
+        self.shouldStop = False
+
+    def stop(self):
+        """Indicate that tests should stop running."""
+        self.shouldStop = True
+
+
 class MultiTestResult(TestResult):
     """A test result that dispatches to many test results."""
 
diff --git a/testtools/tests/test_testresult.py b/testtools/tests/test_testresult.py
index 1b1f656..ac31635 100644
--- a/testtools/tests/test_testresult.py
+++ b/testtools/tests/test_testresult.py
@@ -22,11 +22,13 @@ from testtools import (
     ExtendedToOriginalDecorator,
     MultiTestResult,
     PlaceHolder,
+    StreamFailfast,
     StreamResult,
     StreamSummary,
     StreamToDict,
     Tagger,
     TestCase,
+    TestControl,
     TestResult,
     TestResultDecorator,
     TestByTestResult,
@@ -557,6 +559,12 @@ class TestStreamToDictContract(TestCase, TestStreamResultContract):
         return StreamToDict(lambda x:None)
 
 
+class TestStreamFailfastContract(TestCase, TestStreamResultContract):
+
+    def _make_result(self):
+        return StreamFailfast(lambda:None)
+
+
 class TestDoubleStreamResultEvents(TestCase):
 
     def test_startTestRun(self):
@@ -700,6 +708,47 @@ class TestStreamToDict(TestCase):
             "application/octet-stream", repr(details['another file'].content_type))
 
 
+class TestStreamFailfast(TestCase):
+
+    def test_inprogress(self):
+        result = StreamFailfast(self.fail)
+        result.status('foo', 'inprogress')
+
+    def test_exists(self):
+        result = StreamFailfast(self.fail)
+        result.status('foo', 'exists')
+
+    def test_xfail(self):
+        result = StreamFailfast(self.fail)
+        result.status('foo', 'xfail')
+
+    def test_uxsuccess(self):
+        calls = []
+        def hook():
+            calls.append("called")
+        result = StreamFailfast(hook)
+        result.status('foo', 'uxsuccess')
+        result.status('foo', 'uxsuccess')
+        self.assertEqual(['called', 'called'], calls)
+
+    def test_success(self):
+        result = StreamFailfast(self.fail)
+        result.status('foo', 'success')
+
+    def test_fail(self):
+        calls = []
+        def hook():
+            calls.append("called")
+        result = StreamFailfast(hook)
+        result.status('foo', 'fail')
+        result.status('foo', 'fail')
+        self.assertEqual(['called', 'called'], calls)
+
+    def test_skip(self):
+        result = StreamFailfast(self.fail)
+        result.status('foo', 'skip')
+
+
 class TestStreamSummary(TestCase):
 
     def test_attributes(self):
@@ -844,6 +893,17 @@ testtools.matchers._impl.MismatchError: Differences: [
         self.assertEqual("foo.bar", result.unexpectedSuccesses[0].id())
 
 
+class TestTestControl(TestCase):
+
+    def test_default(self):
+        self.assertEqual(False, TestControl().shouldStop)
+
+    def test_stop(self):
+        control = TestControl()
+        control.stop()
+        self.assertEqual(True, control.shouldStop)
+
+
 class TestTestResult(TestCase):
     """Tests for 'TestResult'."""
